### DISCUSSION TITLE
RAW socket over XFRM interface

### DISCUSSION THREAD

**Discussion by cyanideangel2000** [strongswan/strongswan#1445] [Q&A] [answered]
hello Team,

Wanted to check if this has been already tried.
Can we use RAW socket to directly read the cleartext traffic post xfrm decryption form the xfrm interface?
Further, can i write the cleartext directly to xfrm interface using the same raw socket, will xfrmi run these packets through xfrm encryption and send out?

Thank you.

**Comment by tobiasbrunner** [comment]
Did you try? Anyway, not sure what exactly you mean with "directly" in relation to raw sockets and XFRM interfaces? As opposed to what?

**Reply by tewarivarun** [reply]
Didn't try it yet, but am going to.
I will keep the thread posted with the results from my test.

**Comment by Thermi** [comment]
XFRM interfaces are a kernel feature, not done by userspace so presumably it works. the `xfrmi`tool is only required for setup and teardown (during runtime; like any linux network interface it's just gone when the system shuts down/the kernel stops execution)

**Reply by cyanideangel2000** [reply]
To update, i gave this a quick run today.
I tried to write an UDP packet (matching the xfrm policy outbound) on xfmi using an AF_INET SOCK_RAW socket and it didn't work.
The interesting thing being, all calls were successful including sendto, but didn't see any packet on the xfrmi when monitoring using tcpdump or didn't see the counters of SA increasing even though the code was continuously writing data successfully via the raw socket.

**Reply by tobiasbrunner** [reply]
> on xfmi using an AF_INET SOCK_RAW socket

What does that mean? How do you write "on xfrmi" with a raw socket? Are the packets sent on the physical interface? Or what happens with them?

**Comment by cyanideangel2000** [comment, marked as answer]
Tried with AF_PACKET SOCK_RAW and created an IP packet no ethernet header matching outgoing xfrm policy, and things does seem to work.
So yes, as we expect the case to be, we can use packets sockets to read/write from the xfrmi.
